v2/yarpcrouter/router.go

```diff
@@ -58,19 +58,43 @@ type MapRouter struct {
 func NewMapRouter(defaultService string) MapRouter {
 	return MapRouter{
 		defaultService:            defaultService,
-		serviceProcedureEncodings: make(map[serviceProcedureEncoding]yarpc.Procedure),
+		serviceProcedureEncodings: make(map[serviceProcedureEncoding]yarpc.TransportProcedure),
 		serviceNames:              map[string]struct{}{defaultService: {}},
 	}
 }
 
+// NewMapRouterWithEncodingProcedures constructs a new MapRouter with the given default service name and registers
+// the given encoding-level procedures as transport-level procedures
+func NewMapRouterWithEncodingProcedures(defaultService string, encodingProcedures []yarpc.EncodingProcedure) MapRouter {
+	router := NewMapRouter(defaultService)
+
+	transportProcedures := []yarpc.TransportProcedure{}
+	for _, p := range encodingProcedures {
+		transportProcedures = append(
+			transportProcedures,
+			yarpc.TransportProcedure{
+				Name:        p.Name,
+				Service:     p.Service,
+				HandlerSpec: yarpc.NewUnaryTransportHandlerSpec(yarpc.UnaryTransportHandlerFunc(p.TransportProcedureFunc)),
```

Comment:
In the end, this line should look like:

```
HandlerSpec: yarpc.NewUnaryTransportHandlerSpec(unaryTransportHandler{p}),
```

This is achievable by moving the implementation of `p.TransportProcedureFunc` into a private type that delegates to the unary encoding procedure.

```
struct unaryTransportHandler struct {
    h yarpc.UnaryEncodingHandler
}

var _ yarpc.UnaryTransportHandler = unaryTransportHandler{}

func (h unaryTransportHandler) Handle(Context, Request, Buffer) (Response, Buffer, error) {
  // ...
}
```

We will also need to switch on RPC type (unary or streaming) so that streaming handlers can be registered too. For now, it is sufficient to verify and panic if it’s not unary. We should make a note in CHANGES if this isn’t finished for streaming, since that blocks a release.